### PR TITLE
feat: wait for pending navigation to resolve before many actions

### PR DESCRIPTION
**Pull request by dgozman**
This includes all actions that perform locator handler check.
    
Note this makes it impossible to interact with the page while a main frame navigation is ongoing. This was already the case for Chromium, but now WebKit and Firefox align with it.

Setting `PLAYWRIGHT_SKIP_NAVIGATION_CHECK` environment variable disables this behavior.

### CODE REVIEW

File: packages/playwright-core/src/server/dom.ts [hunk 1]
```diff
@@ -292,7 +292,7 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
     };
   }
 
-  async _retryAction(progress: Progress, actionName: string, action: (retry: number) => Promise<PerformActionResult>, options: { trial?: boolean, force?: boolean, skipLocatorHandlersCheckpoint?: boolean }): Promise<'error:notconnected' | 'done'> {
+  async _retryAction(progress: Progress, actionName: string, action: (retry: number) => Promise<PerformActionResult>, options: { trial?: boolean, force?: boolean, skipActionPreChecls?: boolean }): Promise<'error:notconnected' | 'done'> {
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/dom.ts:295]:
```suggestion
  async _retryAction(progress: Progress, actionName: string, action: (retry: number) => Promise<PerformActionResult>, options: { trial?: boolean, force?: boolean, skipActionPreChecks?: boolean }): Promise<'error:notconnected' | 'done'> {
```

File: packages/playwright-core/src/server/frames.ts [hunk 2]
```diff
@@ -248,6 +248,11 @@ export class FrameManager {
     const frame = this._frames.get(frameId);
     if (!frame)
       return;
+    const pending = frame.pendingDocument();
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/frames.ts:251]: Looks unrelated, land as a separate change?

File: packages/playwright-core/src/server/page.ts [hunk 3]
```diff
@@ -455,7 +456,31 @@ export class Page extends SdkObject {
     this._locatorHandlers.delete(uid);
   }
 
-  async performLocatorHandlersCheckpoint(progress: Progress) {
+  async performActionPreChecks(progress: Progress) {
+    await this._performWaitForNavigationCheck(progress);
+    progress.throwIfAborted();
+    await this._performLocatorHandlersCheckpoint(progress);
```
Comment (pavelfeldman) [inline packages/playwright-core/src/server/page.ts:462]: What if this causes navigation?

**Review by pavelfeldman — APPROVED** [review]

File: packages/playwright-core/src/server/page.ts (hunk 3 above)
Comment (dgozman) [reply packages/playwright-core/src/server/page.ts:462]: Good call, I'll wait for navigations after this one as well.

**Comment by chrisbottin** [comment]
Hey @dgozman 

We have noticed that some of our tests stopped working after bumping our Playwright version from `1.48.2` to `1.49.0`.

I debugged the issue and I pinned it down to the changes in this pull request.

The problem we are having is when we click on a link which triggers the `beforeunload` dialog, after we dismiss the dialog and try to interact with the page again, we get timeout errors.

Here is a simple test replicating the error ...

```js
const {chromium} = require('playwright');
const {expect} = require('@playwright/test');

(async () => {
    const browser = await chromium.launch();
    const page = await browser.newPage();

    await page.mainFrame().goto('https://www.npmjs.com/');

    expect(await page.mainFrame().title()).toStrictEqual('npm | Home');

    await page.evaluate(() => {
        window.onbeforeunload = () => false;
    });

    page.on('dialog', async dialog => {
        await dialog.dismiss();
    });

    await page.getByRole('menuitem', {name: 'Teams'}).click({noWaitAfter: true});

    await page.evaluate(() => {
        window.onbeforeunload = null;
    });

    // Timeout error will occur here:
    await page.getByRole('menuitem', {name: 'Teams'}).click({timeout: 1000});

    expect(await page.mainFrame().title()).toStrictEqual('npm | Teams');

    await browser.close();
})();
```

The test fails on the line `await page.getByRole('menuitem', {name: 'Teams'}).click({timeout: 1000});` with the error:

```
locator.click: Timeout 1000ms exceeded.
Call log:
  - waiting for getByRole('menuitem', { name: 'Teams' })
  - waiting for navigation to finish...
```

During debugging, I noticed the `pendingDocument()` is set to `{documentId: undefined, request: undefined}`

If this line https://github.com/microsoft/playwright/blob/b5bd543cc6935f62732da3b6183bd319eb2bdc2a/packages/playwright-core/src/server/page.ts#L472 is changed to the following, our problem disappears:

```js
if (!mainFrame || !mainFrame.pendingDocument() || !mainFrame.pendingDocument().documentId)
      return;
```

Is this a genuine issue that should be fixed?

**Comment by dgozman** [comment]
@chrisbottin This is a genuine issue, I filed #33806 on your behalf. We'll take a look.

**Comment by chrisbottin** [comment]
@dgozman Thanks, I also raised https://github.com/microsoft/playwright/issues/33803 after posting my comment here.
I've closed it as duplicate now.
